File: kpp/kpp-bench/kpp.js
#!/usr/bin/env node
fs = require("fs")
seedrandom = require("seedrandom")
Math.random = seedrandom(1234567891)

the = { p: 2, Few: 32, Stop: 4, file: "big.csv" }

let rows = [],
    lo = {},
    hi = {},
    xcols = {},
    names = []

atom  = x => isNaN(x = +x) ? (x == "true" ? true : x == "false" ? false : x) : x
split = s => s.split(",").map(atom)
sum   = t => t.reduce((a, b) => a + b, 0)
any   = t => t[Math.floor(Math.random() * t.length)]
norm  = (c, x) => (x - lo[c]) / ((hi[c] - lo[c]) + 1e-32)

xdist = (r1, r2) => {
  let d = 0, n = 0
  for (let c in xcols) {
    let a = norm(c, r1[c]),
        b = norm(c, r2[c])
    d += Math.abs(a - b) ** the.p
    n++ }
  return (d / n) ** (1 / the.p) }

update = (j, x) => {
  if (xcols[j]) {
    lo[j] = lo[j] === undefined ? x : Math.min(lo[j], x)
    hi[j] = hi[j] === undefined ? x : Math.max(hi[j], x) }
  return x }

readHeader = hdr =>
  hdr.map((name, j) => {
    if (!/[+-]$/.test(name)) xcols[j] = true
    return name })

readCSV = file => {
  lines = fs.readFileSync(file, "utf-8").split(/\r?\n/).filter(Boolean)
  names = readHeader(lines[0].split(","))
  rows = lines.slice(1).map(line => split(line).map(update)) }

kpp = (k = the.Stop, r = rows) => {
  let c = [any(r)],
      s = r.slice(0, the.Few)
  for (let i = 1; i < k; i++) {
    let d = s.map(x => Math.min(...c.map(y => xdist(x, y) ** 2)))
    let r = Math.random() * sum(d)
    for (let j = 0; j < d.length; j++) {
      if ((r -= d[j]) <= 0) {
        c.push(s.splice(j, 1)[0])
        break } } }
  return c }

readCSV(the.file)
kpp().forEach(r => {}) // output suppressed for timing

File: kpp/kpp-bench/kpp.py
#!/usr/bin/env python3
import math, random

the = {
    'p': 2,
    'Few': 32,
    'Stop': 4,
    'seed': 1234567891,
    'file': 'big.csv'
}
random.seed(the['seed'])

data = []
lo, hi, xcols, names = {}, {}, {}, []

def atom(x):
    try: return int(x)
    except: pass 
    try: return float(x)
    except: pass
    x = x.strip()
    if x.lower() == "true": return True
    if x.lower() == "false": return False
    return x

def split(line) : return [atom(x) for x in line.strip().split(",")]
def sumf(lst)   : return sum(lst)
def anyof(lst)  : return random.choice(lst)
def norm(c, x)  : return (x - lo[c]) / ((hi[c] - lo[c]) + 1e-32)

def xdist(r1, r2):
    d, n = 0, 0
    for c in xcols:
        a, b = norm(c, r1[c]), norm(c, r2[c])
        d += abs(a - b) ** the['p']
        n += 1
    return (d / n) ** (1 / the['p'])

def update(j, x):
    if j in xcols:
        lo[j] = x if j not in lo else min(lo[j], x)
        hi[j] = x if j not in hi else max(hi[j], x)
    return x

def read_csv(file):
    global names, data
    with open(file) as f:
        lines = [line for line in f if line.strip()]
    names = split(lines[0])
    for j, name in enumerate(names):
        if not name.endswith("+") and not name.endswith("-"):
            xcols[j] = True
    for line in lines[1:]:
        row = [update(j, x) for j, x in enumerate(split(line))]
        data.append(row)

def kpp(k=the['Stop'], rows=None):
    rows = rows or data
    centroids = [anyof(rows)]
    some = rows[:min(the['Few'], len(rows))]
    for _ in range(1, k):
        dists = [min(xdist(x, y) ** 2 for y in centroids) for x in some]
        r = random.random() * sumf(dists)
        for j, d in enumerate(dists):
            r -= d
            if r <= 0:
                centroids.append(some.pop(j))
                break
    return centroids

read_csv(the['file'])
kpp()  # output suppressed for timing

